### PR TITLE
[API][WASM] Add TENSOR_QUANT8_SYMM_PER_CHANNEL and TENSOR_QUANT8_ASYMM_SIGNED

### LINKED ISSUE #1155: [API] support TENSOR_QUANT8_SYMM_PER_CHANNEL
It derives from [`ANEURALNETWORKS_TENSOR_QUANT8_SYMM_PER_CHANNEL`](https://developer.android.com/ndk/reference/struct/a-neural-networks-symm-per-channel-quant-params) of NNAPI which you can use for representing weights to CONV/DEPTHWISE_CONV/TRANSPOSED_CONV operations.

### PR DESCRIPTION
**Pull request by huningxin**
Fix #1155

### CODE REVIEW

**Comment by huningxin** [comment]
Open this PR for collaboration.

@lisa0314 , could you please help check with `TENSOR_QUANT8_ASYMM` and `TENSOR_QUANT8_SYMM_PER_CHANNEL` whether we support this [MobileNet v1 int8 model](https://github.com/intel/optimized-models/tree/v1.0.8/pytorch/mlperf_tools)?

@cuiyanx , could you please help check the Android CTS test for `ANEURALNETWORKS_TENSOR_QUANT8_SYMM_PER_CHANNEL`? We need to convert the native test cases to js to test it.

File: docs/api.md [hunk 1]
```diff
@@ -89,14 +90,18 @@ interface NeuralNetworkContext {
 dictionary OperandOptions {
   required long type;
   sequence<unsigned long> dimensions;
-  // scale: an non-negative floating point value
   float scale;
-  // zeroPoint: an integer, in range [0, 255]
   long zeroPoint;
 };
 
+disctionary OperandSymmPerChannelQuantParams {
```
Comment (lisa0314) [inline docs/api.md]: disctionary => dictionary

**Comment by lisa0314** [comment]
For dnnl_primitive_attr_set_output_scales(https://intel.github.io/mkl-dnn/group__dnnl__api__attributes.html#gae3ff1fe9449f2d9b1cd02ad864b57057) , the set i-th bit for the value of mask indicates that a dedicated output scaling factor is used for each index along that dimension.  ANEURALNETWORKS_TENSOR_QUANT8_SYMM_PER_CHANNEL is one type of mask setting to a particular value. 
However, for the [MobileNet v1 int8 model](https://github.com/intel/optimized-models/tree/v1.0.8/pytorch/mlperf_tools], I suppose it's enough.

**Comment by lisa0314** [comment]
This is a draft using new type definition:
`Let  input0 = {type: nn.TENSOR_QUANT8_ASYMM,
                      Dimensions: [257*3*224*224], scale: 0.007874015718698502,
                      Zero_point: 0};
model.addOperand(input0);
operandIndex++;

Let weights = {type:nn. TENSOR_QUANT8_SYMM_PER_CHANNEL,
                        Dimensions: [32*3*3*3], scale = 0, zero_point = 0}
model.addOperand(weights);
model.setOperandValue((operandIndex, new Uint8Array([2, 2, 2, 2…….]));
model.setOperandSymmPerChannelQuantParams(operandIndex++, { channelDim:0, scales: [0.00013……….]}; // scales size = 32
`

File: docs/api.md (hunk 1 above)
Comment (huningxin) [reply docs/api.md]: good catch. will fix.

Comment (huningxin) [reply docs/api.md]: fixed.

**Comment by huningxin** [comment]
> But the bias for Int8ConvRelu shown as below, how to represent it and its scales?

According to NNAPI, I suppose the bias would use TENSOR_INT32 and its scale would depend on input scale and filter scale, where `bias_scale[i] = input_scale * filter_scale[i]`

**Comment by huningxin** [comment]
> @huningxin For [dnnl_primitive_attr_set_output_scales](https://intel.github.io/mkl-dnn/group__dnnl__api__attributes.html#gae3ff1fe9449f2d9b1cd02ad864b57057) , the set i-th bit for the value of mask indicates that a dedicated output scaling factor is used for each index along that dimension. ANEURALNETWORKS_TENSOR_QUANT8_SYMM_PER_CHANNEL is one type of mask setting to a particular value.

According to NNAPI, when using ANEURALNETWORKS_TENSOR_QUANT8_SYMM_PER_CHANNEL, for CONV_2D filter, the channel dimension (ANeuralNetworksSymmPerChannelQuantParams::channelDim) must be set to 0. For DEPTHWISE_CONV_2D, the channel dimension must be set to 3.

For DNNL `dnnl_primitive_attr_set_output_scales`, I suppose the mask should be 1, as
```
const int mask = 0
    | (1 << 1); // scale per OC dimension, which is the dim #1 on dst tensor:
                // (BATCH, OC, OH, OW)
                //    0     1   2   3
```

**Comment by huningxin** [comment]
> This is a draft for Int8ConvRelu using new type definition:
> `Let input0 = {type: nn.TENSOR_QUANT8_ASYMM, Dimensions: [257*3*224*224], scale: 0.007874015718698502, Zero_point: 0};`

In MobileNet v1 int8 model, I found `Conv__224` has attribute `X_zero_point` is 128, do we need to set `zero_point: 128` here? Does DNNL support zero point? I found [](https://intel.github.io/mkl-dnn/group__dnnl__api__attributes.html#ga6f4b625c7f238883fe32cbf6829b7f30).

> `model.addOperand(input0);`
> `operandIndex++;`
> 
> `Let weights = {type:nn. TENSOR_QUANT8_SYMM_PER_CHANNEL, Dimensions: [32*3*3*3], scale = 0, zero_point = 0};`

For `TENSOR_QUANT8_SYMM_PER_CHANNEL`, we don't need to set `scale` and `zero_point`.

> `model.addOperand(weights);`
> `model.setOperandValue((operandIndex, new Uint8Array([2, 2, 2, 2…….]));`

Should it be `Int8Array`?

> `model.setOperandSymmPerChannelQuantParams(operandIndex++, { channelDim:0, scales: [0.00013……….]}; // scales size = 32`

It should use `Float32Array`, e.g. `scale: new Float32Array([0.00013....])`.

**Comment by huningxin** [comment]
> @cuiyanx , could you please help check the Android CTS test for `ANEURALNETWORKS_TENSOR_QUANT8_SYMM_PER_CHANNEL`? We need to convert the native test cases to js to test it.

ping @cuiyanx .

**Comment by lisa0314** [comment]
According to NNAPI, when using ANEURALNETWORKS_TENSOR_QUANT8_SYMM_PER_CHANNEL, for CONV_2D filter, the channel dimension (ANeuralNetworksSymmPerChannelQuantParams::channelDim) must be set to 0. For DEPTHWISE_CONV_2D, the channel dimension must be set to 3.

For DNNL `dnnl_primitive_attr_set_output_scales`, I suppose the mask should be 1, as
```
const int mask = 0
    | (1 << 1); // scale per OC dimension, which is the dim #1 on dst tensor:
                // (BATCH, OC, OH, OW)
                //    0     1   2   3
```
I suppose the mask should be 2 for DNNL `dnnl_primitive_attr_set_output_scales` after `<<`.

**Comment by cuiyanx** [comment]
@huningxin @lisa0314 
I update [CTS converter](https://github.com/cuiyanx/CTSConverter/tree/master/src/select_per_channel) and some [CTS test cases](https://github.com/cuiyanx/dnnl-models/tree/master/select-CTS-test/per-channel) to support `ANEURALNETWORKS_TENSOR_QUANT8_SYMM_PER_CHANNEL` op.
Please take a look.

**Comment by huningxin** [comment]
> I suppose the mask should be 2 for DNNL `dnnl_primitive_attr_set_output_scales` after `<<`.
> A mask = 2 = 2^1 selects the output channel for scaling.

You are right @lisa0314 , the mask should be 2.

**Comment by huningxin** [comment]
Thanks @cuiyanx .

Could you please share with us where to find the corresponding native NNAPI test cases for `ANEURALNETWORKS_TENSOR_QUANT8_SYMM_PER_CHANNEL`? So we can compare the difference.

**Comment by huningxin** [comment]
> I update [CTS converter](https://github.com/cuiyanx/CTSConverter/tree/master/src/select_per_channel) and some [CTS test cases](https://github.com/cuiyanx/dnnl-models/tree/master/select-CTS-test/per-channel) to support

Hi @cuiyanx , could you please open a PR to add these CTS test into repo? With that, it would be easier to collaborate on the implementation and test cases.

**Comment by cuiyanx** [comment]
@huningxin @lisa0314
I have add NNAPI CTS test cases to [select-CTS-test](https://github.com/cuiyanx/dnnl-models/tree/master/select-CTS-test/per-channel/V1_2).

**_Note_**:
`xxx.js`: generated test case by our CTS converter.
`xxx.mod.py`: native NNAPI test case.
`xxx.mod.py.cpp`: generated test case by native CTS converter.
`xxx.model.cpp`: generated model file.
`xxx.example.cpp`: generated example file.

**Comment by huningxin** [comment]
Regression test result: passes: 482 failures: 255 (baseline passes: 482 failures: 255), no regression.

Passed test cases added by https://github.com/intel/webml-polyfill/pull/1162 and https://github.com/intel/webml-polyfill/pull/1180

@lisa0314 @cuiyanx , please take a look at API change.

@akineeic , please take a look at WASM polyfill change.

**Comment by akineeic** [comment]
> Regression test result: passes: 482 failures: 255 (baseline passes: 482 failures: 255), no regression.
> 
> Passed test cases added by #1162 and #1180
> 
> @lisa0314 @cuiyanx , please take a look at API change.
> 
> @akineeic , please take a look at WASM polyfill change.

LGTM, thanks @huningxin !

**Comment by cuiyanx** [comment]
@huningxin LGTM, thanks!

**Comment by huningxin** [comment]
Thanks @cuiyanx and @akineeic for review. Let's merge it.

**Pull request merged**